query/src/main/java/io/keen/client/java/KeenQueryClient.java

```diff
@@ -0,0 +1,704 @@
+        if (input instanceof Integer) {
+            Integer intValue = (Integer) input;
+            thisObject = new LongResult(intValue.longValue());
+        } else if (input instanceof Long) {
+            thisObject = new LongResult((Long) input);
+        } else if (input instanceof Double) {
+            thisObject = new DoubleResult((Double) input);
+        } else if (input instanceof String) {
+            thisObject = new StringResult((String) input);
+        } else if (input instanceof ArrayList) {
+
+            // recursively construct the children of this...
+            ArrayList<QueryResult> listOutput = new ArrayList<QueryResult>();
+            ArrayList<Object> listInput = (ArrayList<Object>)input;
+
+            // if this is an IntervalResult, construct the IntervalResult object.
+            if (isInterval) {
+                thisObject = constructIntervalResult(listInput, isGroupBy);
+            } else if (isGroupBy) {
+                // if this is a GroupByResult, construct the GroupByResult object.
+                // Note that if this is both an Interval and GroupBy, the GroupBy
+                // code will be called later from within constructIntervalResult()
+                thisObject = constructGroupByResult(listInput);
+            } else {
+
+                // else if this is just a List of QueryResult objects - for example,
+                // Select Unique query returns a list of unique objects.
+                for (Object child : listInput) {
+                    QueryResult resultItem = constructQueryResult(child, false, false);
+                    listOutput.add(resultItem);
+                }
+                thisObject = new ListResult(listOutput);
+            }
+        }
+
+        return thisObject;
+    }
+
+    private static IntervalResult constructIntervalResult(ArrayList<Object> intervals, boolean isGroupBy) {
+        Map<AbsoluteTimeframe, QueryResult> intervalResult = new HashMap<AbsoluteTimeframe, QueryResult>();
+
+        for (Object child : intervals) {
+            if (child instanceof HashMap) {
+                HashMap<String, Object> inputMap = (HashMap<String, Object>) child;
+                // If this is an interval, it should have keys "timeframe" and "value"
+                if (inputMap.containsKey(KeenQueryConstants.TIMEFRAME) && (inputMap.containsKey(KeenQueryConstants.VALUE))) {
+                    AbsoluteTimeframe absoluteTimeframe = null;
+                    Object timeframe = inputMap.get(KeenQueryConstants.TIMEFRAME);
+                    if (timeframe instanceof HashMap) {
+                        HashMap<String, String> hashTimeframe = (HashMap<String, String>) timeframe;
+                        String start = hashTimeframe.get(KeenQueryConstants.START);
+                        String end = hashTimeframe.get(KeenQueryConstants.END);
+                        absoluteTimeframe = new AbsoluteTimeframe(start, end);
+                    }
```

Comment:
Concern: What happens if `timeframe` is not a `HashMap`? Seems like that case should be handled as an error; in the current code you'd just end up with putting a `null` key into the result map (which might even throw a NullPointerException?).